lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -401,6 +457,252 @@ LogicalResult ConcatRefOp::inferReturnTypes(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// StructCreateOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult StructCreateOp::verify() {
+  /// checks if the types of the inputs are exactly equal to the types of the
+  /// result struct fields
+  return TypeSwitch<Type, LogicalResult>(getType().getNestedType())
+      .Case<StructType, UnpackedStructType>([this](auto &type) {
+        auto members = type.getMembers();
+        auto inputs = getInput();
+        if (inputs.size() != members.size())
+          return failure();
+        for (size_t i = 0; i < members.size(); i++) {
+          auto memberType = cast<UnpackedType>(members[i].type);
+          auto inputType = cast<RefType>(inputs[i].getType()).getNestedType();
+          if (inputType != memberType) {
+            emitOpError("input types must match struct field types and orders");
+            return failure();
+          }
+        }
+        return success();
+      })
+      .Default([this](auto &) {
+        emitOpError("Result type must be StructType or UnpackedStructType");
+        return failure();
+      });
+}
+
+//===----------------------------------------------------------------------===//
+// StructExtractOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult StructExtractOp::verify() {
+  /// checks if the type of the result match field type in this struct
+  return TypeSwitch<Type, LogicalResult>(getInput().getType())
+      .Case<StructType, UnpackedStructType>([this](auto &type) {
+        auto members = type.getMembers();
+        auto filedName = getFieldName();
+        auto resultType = getType();
+        for (const auto &member : members) {
+          if (member.name == filedName && member.type == resultType) {
+            return success();
+          }
+        }
+        emitOpError("result type must match struct field type");
+        return failure();
+      })
+      .Default([this](auto &) {
+        emitOpError("input type must be StructType or UnpackedStructType");
+        return failure();
+      });
+}
+
+bool StructExtractOp::canRewire(const DestructurableMemorySlot &slot,
+                                SmallPtrSetImpl<Attribute> &usedIndices,
+                                SmallVectorImpl<MemorySlot> &mustBeSafelyUsed,
+                                const DataLayout &dataLayout) {
+  return TypeSwitch<Type, bool>(getType())
+      .Case<StructType, UnpackedStructType>(
+          [this, &slot, &usedIndices](auto &type) {
+            if (slot.ptr != getInput())
+              return false;
+            auto index = getFieldNameAttr();
+            if (!index)
+              return false;
+            usedIndices.insert(index);
+            return true;
+          })
+      .Default([](auto) { return false; });
+}
```

Comment:
I assume all struct field references are safely used by design, right? (from a reference to a field of a struct, one cannot access other fields of the same struct)

Maybe add a comment about this if it is correct.

Reply:
Yes you are right, it is safe as far as I know

Reply:
Okay good. I was just asking because if it is not safe you need to add it to the safely used check so I was not sure if it was intended or forgotten.

Reply:
I did notice memrefOp did some mutable-clear operation. I think the point is its Ops index several fields, but this Op only index one field.

Reply:
Yes, I was more thinking about LLVM-like use cases where pointers point to somewhere in the bit byte array that is memory. All good.